pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -929,6 +929,16 @@ func combineTaskRunAndTaskSpecLabels(pr *v1beta1.PipelineRun, pipelineTask *v1be
 			labels[key] = value
 		}
 	}
+
+	// check if a task is part of the finally section, add a label to identify it during the runtime
+	if pr.Status.PipelineSpec != nil {
+		for _, f := range pr.Status.PipelineSpec.Finally {
+			if pipelineTask.Name == f.Name {
+				labels[pipeline.GroupName+pipeline.FinallyLabelKey] = "true"
+				break
+			}
+		}
+	}
```

Comment:
This doesn't feel like 100% right place for this, this function combines `Task` spec and `TaskRun` labels, which is why the function does not have access to the pipeline spec, and you have to go and get it from the status. Whether a `PipelineTask` belongs to the `finally` section is a `Pipeline` level concern, so perhaps we could pass that info down to the `createTaskRun` function, adding an extra boolean param:

```
				if rprt.IsFinalTask(pipelineRunFacts) {
					rprt.TaskRun, err = c.createTaskRun(ctx, rprt, pr, as.StorageBasePath(pr), getFinallyTaskRunTimeout, True)
				} else {
					rprt.TaskRun, err = c.createTaskRun(ctx, rprt, pr, as.StorageBasePath(pr), getTaskRunTimeout, False)
				}
```

Alternatively you could use the [`GetPipelineFunc`](https://github.com/tektoncd/pipeline/blob/a728ce3c3d717fe52ff0b1eb71d357460498115c/pkg/reconciler/pipelinerun/resources/pipelineref.go#L36) function, which returns the spec from the status if available, or it will fetch it otherwise.

Reply:
thanks @afrittoli I think this was a reasonable place since its combining labels from the `task` and `taskRun`. Thanks for suggestion, I have moved it to `getTaskrunLabels` function where we are adding other `tekton.dev` labels such as `tekton.dev/pipelineRun` and `tekton.dev/pipelineTask`.

Also, its safe and reasonable to get the specification from the `status` after we added it as a source of truth in https://github.com/tektoncd/pipeline/pull/3941

`createTaskRun` is growing complex with every new functionality we add. We really need a `factory` style interface for the `taskRun` and `Run` instead of operating on the parameters.